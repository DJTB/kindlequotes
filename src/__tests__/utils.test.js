const { single, multiple, withDuplicates, withBookmarks, alternateFormat } = require('./fixtures');

const {
  contractSpaces,
  isEmpty,
  monthNameToNum,
  parseAuthors,
  parseContent,
  parseDate,
  parseLoc,
  parseTitle,
  prependEllipsis,
  prependZero,
  reorderNames,
  match,
  smartQuotes,
  stripBOM,
  swapJoin,
  to24Hr,
  transformQuotes,
} = require('../utils');

describe('stripBom()', () => {
  it('has a sane default', () => {
    expect(stripBOM()).toBe('');
  });
});

describe('monthNameToNum()', () => {
  it('has a sane default', () => {
    expect(monthNameToNum()).toBe(-1);
  });
  it('works as expected', () => {
    const expected = [
      'January',
      'February',
      'March',
      'April',
      'May',
      'June',
      'July',
      'August',
      'September',
      'October',
      'November',
      'December',
    ];
    const actual = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
    expect(expected.map(monthNameToNum)).toEqual(expect.arrayContaining(actual));
  });
});

describe('match()', () => {
  it('has a sane default', () => {
    expect(match()).toEqual([]);
  });
  it('returns matches', () => {
    expect(match('anapple', /(an)(apple)/)).toEqual(
      expect.arrayContaining(['anapple', 'an', 'apple'])
    );
  });
  it('returns empty array instead of null if no match', () => {
    expect(match('anapple', /oranges/)).toEqual([]);
  });
});

describe('contractSpaces()', () => {
  it('sane default', () => {
    expect(contractSpaces()).toBe('');
  });
  it('works as expected', () => {
    expect(contractSpaces('          ')).toBe(' ');
  });
});

describe('isEmpty()', () => {
  it('sane default', () => {
    expect(isEmpty()).toBe(true);
  });
  it('works as expected', () => {
    expect(isEmpty('')).toBe(true);
    expect(isEmpty([])).toBe(true);
    expect(isEmpty(null)).toBe(true);
  });
});

describe('prependZero()', () => {
  it('sane default', () => {
    expect(prependZero()).toBe('');
  });
  it('works as expected', () => {
    expect(prependZero(0)).toBe('00');
    expect(prependZero(1)).toBe('01');
    expect(prependZero(10)).toBe('10');
  });
});

describe('prependEllipsis()', () => {
  it('sane default', () => {
    expect(prependEllipsis()).toBe('');
  });
  it('works as expected', () => {
    expect(prependEllipsis('Starting sentence')).toBe('Starting sentence');
    expect(prependEllipsis('mid sentence')).toBe('…mid sentence');
  });
});

describe('reorderNames()', () => {
  it('sane default', () => {
    expect(reorderNames()).toBe('');
  });
  it('works as expected', () => {
    expect(reorderNames('First Last')).toBe('First Last');
    expect(reorderNames('Last, First')).toBe('First Last');
  });
});

describe('smartQuotes()', () => {
  it('sane default', () => {
    expect(smartQuotes()).toBe('');
  });
  it('works as expected', () => {
    expect(smartQuotes('"He said \'wow man\'"')).toBe('“He said ‘wow man’”');
  });
});

describe('swapJoin()', () => {
  it('sane default', () => {
    expect(swapJoin()).toBe('');
  });
  it('works as expected', () => {
    expect(swapJoin('one', 'two')).toBe('two one');
  });
});

describe('parseTitle()', () => {
  it('sane default', () => {
    expect(parseTitle()).toBe('');
  });
  it('works as expected', () => {
    expect(parseTitle('Acceptance: A Novel (The Southern Reach Trilogy) (Jeff VanderMeer)')).toBe(
      'Acceptance: A Novel (The Southern Reach Trilogy)'
    );
  });
});

describe('parseAuthors()', () => {
  it('sane default', () => {
    expect(parseAuthors()).toEqual([]);
  });
  it('works as expected', () => {
    expect(
      parseAuthors('Acceptance: A Novel (The Southern Reach Trilogy) (Jeff VanderMeer)')
    ).toEqual(expect.arrayContaining(['Jeff VanderMeer']));
    expect(parseAuthors('Songs of the Dying Earth (Dozois, Gardner;Martin, George R.R.)')).toEqual(
      expect.arrayContaining(['Gardner Dozois', 'George R.R. Martin'])
    );
  });
});

describe('parseDate()', () => {
  it('sane default', () => {
    expect(parseDate()).toBe('');
  });
  it('works as expected', () => {
    expect(
      parseDate(
        '- Your Highlight at location 11849-11850 | Added on Thursday, 30 April 2015 12:58:20'
      )
    ).toEqual('2015-04-30T12:58:20.000Z');
  });
});

describe('parseLoc()', () => {
  it('sane default', () => {
    expect(parseLoc()).toBe('');
  });
  it('works as expected', () => {
    expect(
      parseLoc(
        '- Your Highlight at location 11849-11850 | Added on Thursday, 30 April 2015 20:58:20'
      )
    ).toBe('11849-11850');
  });
});

describe('to24Hr()', () => {
  it('sane default', () => {
    expect(to24Hr()).toBe('');
  });
  it('works as expected', () => {
    expect(to24Hr('1:10:10 PM')).toBe('13:10:10');
    expect(to24Hr('11:10:10 PM')).toBe('23:10:10');
    expect(to24Hr('1:10:10 AM')).toBe('1:10:10');
    expect(to24Hr('11:10:10 AM')).toBe('11:10:10');
    expect(to24Hr('12:10:10 AM')).toBe('12:10:10');
    expect(to24Hr('12:10:10 PM')).toBe('24:10:10');
  });
});

describe('parseContent()', () => {
  it('sane default', () => {
    expect(parseContent()).toBe('');
  });
  it('works as expected', () => {
    expect(
      parseContent(
        'Blake says, "Where?" He\'s a man who measures words as if he had only a few given to him by Fate; too generous a syllable from his lips, and he might fall over dead.'
      )
    ).toBe(
      'Blake says, “Where?” He’s a man who measures words as if he had only a few given to him by Fate; too generous a syllable from his lips, and he might fall over dead.'
    );

    expect(
      parseContent(
        'and eyeing the wizard speculatively across the room. A glance was enough to tell Molloqos that she was a woman of the evening, though in her case evening was edging on toward night.'
      )
    ).toBe(
      '…and eyeing the wizard speculatively across the room. A glance was enough to tell Molloqos that she was a woman of the evening, though in her case evening was edging on toward night.'
    );
  });
});

describe('transformQuotes()', () => {
  it('sane default', () => {
    expect(transformQuotes()).toEqual([]);
  });
  it('works for a single entry', () => {
    expect(transformQuotes(single)).toMatchSnapshot();
  });
  it('works for multiple entries', () => {
    expect(transformQuotes(multiple)).toMatchSnapshot();
  });
  it('removes duplicates', () => {
    expect(transformQuotes(withDuplicates)).toMatchSnapshot();
  });
  it('ignores bookmarks', () => {
    expect(transformQuotes(withBookmarks)).toMatchSnapshot();
  });
  it('handles alternate format', () => {
    expect(transformQuotes(alternateFormat)).toMatchSnapshot();
  });
});
