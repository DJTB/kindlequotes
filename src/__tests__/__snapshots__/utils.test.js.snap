// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformQuotes() ignores bookmarks 1`] = `
Array [
  Object {
    "author": "Peter Watts",
    "content": "“Damn right. And you’ll be talking to ’em while I’m giving them their physicals.” “Maybe not. I mean, Mom would never admit it in a million years but you had a point about language. When you get right down to it, it’s a workaround. Like trying to describe dreams with smoke signals. It’s noble, it’s maybe the most noble thing a body can do but you can’t turn a sunset into a string of grunts without losing something. It’s limiting. Maybe whatever’s out here doesn’t even use it.”",
    "date": "2016-08-14T18:04:12.000Z",
    "loc": "1083-1086",
    "title": "Blindsight",
  },
  Object {
    "author": "Peter Watts",
    "content": "None of it had been real, as it turned out. Or at least, none of it had been confirmed. The experts of the day had been little more than witch doctors dancing through improvised rituals: meandering free-form interviews full of leading questions and nonverbal cues, scavenger hunts through regurgitated childhoods. Sometimes a shot of lithium or haloperidol when the beads and rattles didn’t work. The technology to map minds was barely off the ground; the technology to edit them was years away. So the therapists and psychiatrists poked at their victims and invented names for things they didn’t understand, and argued over the shrines of Freud and Klein and the old Astrologers. Doing their very best to sound like practitioners of Science.",
    "date": "2016-08-15T23:44:11.000Z",
    "loc": "2030-2035",
    "title": "Blindsight",
  },
]
`;

exports[`transformQuotes() removes duplicates 1`] = `
Array [
  Object {
    "author": "Nick Harkaway",
    "content": "The war came to an end and – as with the one before it – everyone was too exhausted to cheer. People smiled at one another not like victorious heroes but like punch-drunk prizefighters, eyes swollen shut and lips rippled and split, who do not understand why everyone is clapping.",
    "date": "2014-07-29T18:39:52.000Z",
    "loc": "6551-6553",
    "title": "Angelmaker",
  },
  Object {
    "author": "Nick Harkaway",
    "content": "‘Home-made,’ Edie Banister says happily, a soon-to-be-novagenarian with a bad attitude and a fine knowledge of exothermic reactions. ‘I think I may have over-egged the nitro and gone a bit heavy on the toluene. However, nothing succeeds like excess, ey?’",
    "date": "2014-08-01T00:25:28.000Z",
    "loc": "7692-7694",
    "title": "Angelmaker",
  },
  Object {
    "author": "Jeff VanderMeer",
    "content": "…had been falling apart for many years. The woods were dark and loamy and took the sound of our transport like a wolf taking a rabbit.",
    "date": "2014-08-31T20:16:53.000Z",
    "loc": "1653-1654",
    "title": "The Third Bear",
  },
  Object {
    "author": "Jeff VanderMeer",
    "content": "Blake says, “Where?” He’s a man who measures words as if he had only a few given to him by Fate; too generous a syllable from his lips, and he might fall over dead.",
    "date": "2014-09-01T12:58:20.000Z",
    "loc": "1856-1857",
    "title": "The Third Bear",
  },
]
`;

exports[`transformQuotes() works for a single entry 1`] = `
Array [
  Object {
    "author": "Jeff VanderMeer",
    "content": "Blake says, “Where?” He’s a man who measures words as if he had only a few given to him by Fate; too generous a syllable from his lips, and he might fall over dead.",
    "date": "2014-09-01T12:58:20.000Z",
    "loc": "1856-1857",
    "title": "The Third Bear",
  },
]
`;

exports[`transformQuotes() works for multiple entries 1`] = `
Array [
  Object {
    "author": "Nick Harkaway",
    "content": "Granted, it belongs to the man to whom everyone else is apparently deferring. He is richly dressed, and surrounded by dependants, concubines, and offspring. And yet, Edie has seen other pictures in the past where one child, caught by chance in an attitude of casual joy, has completely outshone such a parent; where one unthinking scullery maid has glanced at the camera and displayed for a moment her natural beauty, and the social order has been quite overturned. Photography is without mercy – though it’s nonsense to say it does not lie. Rather, it lies in a particular, capricious way which makes beggars of ministers and gods of cat’s meat men.",
    "date": "2014-07-22T23:51:15.000Z",
    "loc": "3023-3027",
    "title": "Angelmaker",
  },
  Object {
    "author": "Nick Harkaway",
    "content": "…and goes upstairs to the relative tranquillity of her rented room. She stares around her. Oscar Wilde, she recalls, acknowledged the close of his mortal existence by remarking: ‘It’s me or the wallpaper. One of us has got to go.’ Looking up at the brown floral print, she feels a fleeting kinship.",
    "date": "2014-07-23T00:22:32.000Z",
    "loc": "3211-3214",
    "title": "Angelmaker",
  },
  Object {
    "author": "Nick Harkaway",
    "content": "The war came to an end and – as with the one before it – everyone was too exhausted to cheer. People smiled at one another not like victorious heroes but like punch-drunk prizefighters, eyes swollen shut and lips rippled and split, who do not understand why everyone is clapping.",
    "date": "2014-07-29T18:39:52.000Z",
    "loc": "6551-6553",
    "title": "Angelmaker",
  },
  Object {
    "author": "Nick Harkaway",
    "content": "‘Home-made,’ Edie Banister says happily, a soon-to-be-novagenarian with a bad attitude and a fine knowledge of exothermic reactions. ‘I think I may have over-egged the nitro and gone a bit heavy on the toluene. However, nothing succeeds like excess, ey?’",
    "date": "2014-08-01T00:25:28.000Z",
    "loc": "7692-7694",
    "title": "Angelmaker",
  },
  Object {
    "author": "Jeff VanderMeer",
    "content": "…had been falling apart for many years. The woods were dark and loamy and took the sound of our transport like a wolf taking a rabbit.",
    "date": "2014-08-31T20:16:53.000Z",
    "loc": "1653-1654",
    "title": "The Third Bear",
  },
  Object {
    "author": "Jeff VanderMeer",
    "content": "Blake says, “Where?” He’s a man who measures words as if he had only a few given to him by Fate; too generous a syllable from his lips, and he might fall over dead.",
    "date": "2014-09-01T12:58:20.000Z",
    "loc": "1856-1857",
    "title": "The Third Bear",
  },
]
`;
